File: frontend/src/app/models/baseVM.ts
import { HttpStatusCode } from "@angular/common/http";

export class ResponseMessage {
    message: string;
    description: string;

    constructor(message: string, description: string) {
        this.message = message;
        this.description = description;
    }
}

export class BaseVM {
    status: HttpStatusCode;
    messages: ResponseMessage[];

    constructor(code: HttpStatusCode = HttpStatusCode.Ok, ...messages: ResponseMessage[]) {
        this.status = code;
        this.messages = messages.length ? messages : [];
    }
}

export class BaseTVM<T> extends BaseVM {
    data: T | null;

    constructor(code: HttpStatusCode = HttpStatusCode.Ok, ...messages: ResponseMessage[]) {
        super(code, ...messages);
        this.data = null; // Initialize data as null
    }
}

File: UserManagementSystem.Application/Dtos/BaseVM.cs
using Newtonsoft.Json;
using System.Net;
using System.Text.Json.Serialization;

namespace UserManagementSystem.Application.Dtos
{
    public class BaseVM
    {
        [JsonProperty("status")]
        public HttpStatusCode Status { get; set; } = HttpStatusCode.OK;

        [JsonProperty("messages")]
        public List<ResponseMessage> Messages { get; set; } = new List<ResponseMessage>();

        public BaseVM() { }
        public BaseVM(HttpStatusCode code, params ResponseMessage[] messages)
        {
            this.Status = code;
            this.Messages = messages?.ToList() ?? new List<ResponseMessage>();
        }
    }
    public class BaseVM<T> : BaseVM
    {
        public T Data { get; set; }
        public BaseVM() : base() { }
        public BaseVM(HttpStatusCode code, params ResponseMessage[] messages) : base(code, messages) { }
    }

    public class ResponseMessage
    {
        [JsonProperty("message")]
        public string Message { get; set; }

        [JsonProperty("description")]
        public string Description { get; set; }
    }
}
